bundles/org.palladiosimulator.dataflow.confidentiality.analysis/src/org/palladiosimulator/dataflow/confidentiality/analysis/sequence/entity/pcm/SEFFActionSequenceElement.java

```diff
@@ -15,10 +29,55 @@ public SEFFActionSequenceElement(T element, Deque<AssemblyContext> context) {
         // TODO Auto-generated constructor stub
     }
 
+    public SEFFActionSequenceElement(SEFFActionSequenceElement<T> oldElement, List<DataFlowVariable> dataFlowVariables, List<CharacteristicValue> nodeVariables) {
+        super(oldElement, dataFlowVariables, nodeVariables);
+    }
+
     @Override
     public AbstractActionSequenceElement<T> evaluateDataFlow(List<DataFlowVariable> variables) {
-        // TODO Auto-generated method stub
-        return null;
+    	List<CharacteristicValue> nodeVariables = this.evaluateNodeCharacteristics();
+        List<VariableCharacterisation> dataflowElements = ((SetVariableAction) super.getElement())
+            .getLocalVariableUsages_SetVariableAction()
+            .stream()
+            .flatMap(it -> it.getVariableCharacterisation_VariableUsage()
+                .stream())
+            .toList();
+        CharacteristicsCalculator characteristicsCalculator = new CharacteristicsCalculator(variables, nodeVariables);
+        dataflowElements.stream()
+            .forEach(it -> characteristicsCalculator.evaluate(it));
+        return new SEFFActionSequenceElement<>(this, characteristicsCalculator.getCalculatedCharacteristics(), nodeVariables);
+    }
+    
+    protected List<CharacteristicValue> evaluateNodeCharacteristics() {
```

Comment:
From a quick look it seems that some of the functionality in `SEFFActionSequenceElement` and `UserActionSequenceElement` is duplicate. Maybe you can also simplify the `evaluateDataFlow` and `evaluateNodeCharacteristics` methods by encapsulating common behavior.